### PR TITLE
eye aspect ratio

### PR DESCRIPTION
**Pull request by israeljacob**

### CODE REVIEW

File: alarms/alarms.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import time
+import playsound
+from utils.consts import ALARM_PATH
+
+
+def alarm():
+    time1 = time.time() + 1
```
Comment (dsl96) [inline alarms/alarms.py:7]: make the sound duration as argument to the function with default value

File: blinked_detector.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    """
+    left_width = dist.euclidean(eye_landmarks[1], eye_landmarks[5])
+    right_width = dist.euclidean(eye_landmarks[2], eye_landmarks[4])
+    length = dist.euclidean(eye_landmarks[0], eye_landmarks[3])
+
+    ratio = (left_width + right_width) / (2.0 * length)
+    return ratio
+
+
+def get_gray_image(image: np.ndarray) -> np.ndarray:
```
Comment (dsl96) [inline blinked_detector.py]: change image arg to image_rgb

File: blinked_detector.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    gray_image = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
+    return gray_image
+
+
+def get_face_rects(image: np.ndarray) -> Optional[List[Tuple]]:
+    """
+    Detect faces in an image using a face detection model.
+
+    :param image: A NumPy array representing an input image.
+    :return: A list of tuples representing detected face rectangles
```
Comment (dsl96) [inline blinked_detector.py:52]: its return only the largest face

File: blinked_detector.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+             as (left, top, right, bottom) coordinates.
+    """
+    rects = FACE_DETECTOR(image, 0)
```
Comment (dsl96) [inline blinked_detector.py]: what is the '0'

File: blinked_detector.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    if not rects:
+        return None
+    largest_face = max(rects, key=lambda rect: (rect.right() - rect.left()) * (rect.bottom() - rect.top()))
```
Comment (dsl96) [inline blinked_detector.py:58]: its not look like a simple numpy array like the docs

File: blinked_detector.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    :param largest_face: A list of tuples representing the coordinates of the largest detected face
+                         as (left, top, right, bottom) coordinates.
+    :return: A NumPy array containing facial landmark points (x, y) for the largest detected face.
+    """
+    shape = PREDICTOR(gray_image, largest_face)
+    shape = face_utils.shape_to_np(shape)
+    return shape
+
+
+def extract_eyes_coordinates(image: np.ndarray) -> Optional[Tuple[Any, Any]]:
```
Comment (dsl96) [inline blinked_detector.py]: change Any to numpy.array

File: blinked_detector.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    """
+    Extracts the coordinates of the left and right eyes from a face image.
+
+    :param image: A NumPy array representing a face image.
+    :return: Two NumPy arrays containing the coordinates of the left and right eyes.
+    """
+    gray_image = get_gray_image(image)
+    rect = get_face_rects(gray_image)
```
Comment (dsl96) [inline blinked_detector.py:84]: change "rect" to "face_rect"

File: blinked_detector.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    shape = face_utils.shape_to_np(shape)
+    return shape
+
+
+def extract_eyes_coordinates(image: np.ndarray) -> Optional[Tuple[Any, Any]]:
+    """
+    Extracts the coordinates of the left and right eyes from a face image.
+
+    :param image: A NumPy array representing a face image.
+    :return: Two NumPy arrays containing the coordinates of the left and right eyes.
```
Comment (dsl96) [inline blinked_detector.py:81]: explain the return format

File: blinked_detector.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    shape = get_face_shape(gray_image, rect)
+    return shape[LEFT_START: LEFT_END], shape[RIGHT_START: RIGHT_END]
+
+
+def are_eyes_blinked(image: np.ndarray) -> (bool, np.ndarray):
+    """
+    Determine if the eyes in the given image are blinked.
+
+    :param image: A representation of the image containing the person's eyes.
+    :return: True if the eyes are blinked, False otherwise.
```
Comment (dsl96) [inline blinked_detector.py:96]: its return eye coordinate also

File: blinked_detector.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    """
```
Comment (dsl96) [inline blinked_detector.py:97]: add to doc its consider only the the largest face

File: blinked_detector.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    Get facial landmarks for the largest detected face in a grayscale image.
+
+    :param gray_image: A NumPy array representing a grayscale image containing a face.
+    :param largest_face: A list of tuples representing the coordinates of the largest detected face
+                         as (left, top, right, bottom) coordinates.
+    :return: A NumPy array containing facial landmark points (x, y) for the largest detected face.
+    """
+    shape = PREDICTOR(gray_image, largest_face)
+    shape = face_utils.shape_to_np(shape)
+    return shape
```
Comment (dsl96) [inline blinked_detector.py:73]: dont use the same name  make new one

File: sleep_deector.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+import numpy as np
+from blinked_detector import are_eyes_blinked
+import cv2
+import time
+from imutils.video import VideoStream
+from utils.consts import MAX_BLINKS
+from alarms.alarms import alarm
+import asyncio
+
+cap = cv2.VideoCapture(0)
```
Comment (dsl96) [inline sleep_deector.py]: not in use

File: sleep_deector.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+vs = VideoStream(src=0).start()
+font = cv2.FONT_HERSHEY_SIMPLEX
```
Comment (dsl96) [inline sleep_deector.py]: move font to consts

File: sleep_deector.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def is_sleeping() -> None:
+    counter = 0
```
Comment (dsl96) [inline sleep_deector.py:45]: change to blink_counter

File: sleep_deector.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import imutils
+import numpy as np
+from blinked_detector import are_eyes_blinked
+import cv2
+import time
+from imutils.video import VideoStream
+from utils.consts import MAX_BLINKS
+from alarms.alarms import alarm
+import asyncio
```
Comment (dsl96) [inline sleep_deector.py]: asyncio not in use

File: sleep_deector.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        frame = vs.read()
+        frame = imutils.resize(frame, width=500)
+
+        key = cv2.waitKey(1)
+
+        is_blinked, eye_coordinates = are_eyes_blinked(np.array(frame))
+        if eye_coordinates:
+            leftEyeHull = cv2.convexHull(eye_coordinates[0])
+            rightEyeHull = cv2.convexHull(eye_coordinates[1])
+            cv2.drawContours(frame, [leftEyeHull], -1, (0, 255, 0), 1)
```
Comment (dsl96) [inline sleep_deector.py]: move color to consts (0,255,0)

File: sleep_deector.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            cv2.drawContours(frame, [rightEyeHull], -1, (0, 255, 0), 1)
```
Comment (dsl96) [inline sleep_deector.py]: drawContours in one func pass the contours in list of counurs[[c1],[c2]]

Comment (dsl96) [inline sleep_deector.py]: optional bug if recognize only one eye

File: sleep_deector.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        if is_blinked and time.time() - start > 0.5:
+            start = time.time()
+            counter += 1
+        elif not is_blinked:
+            counter = 0
+        cv2.putText(frame, str(counter), (10, 350), font, 3, (0, 255, 0), 2, cv2.LINE_AA)
```
Comment (dsl96) [inline sleep_deector.py]: all ars to consts (10, 350), font, 3, (0, 255, 0), 2, cv2.LINE_AA

File: sleep_deector.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+        cv2.imshow("Surgeon's image", frame)
+        if counter >= MAX_BLINKS:
+            alarm()
+        if key == ord('q'):
+            break
+
+
+
+cap.release()
```
Comment (dsl96) [inline sleep_deector.py]: cap not in use

File: utils/consts.py [hunk 20]
```diff
@@ -0,0 +1,13 @@
+from imutils import face_utils
+import dlib
+
+PREDICTOR_PATH = 'blink_detector\\shape_predictor_68_face_landmarks.dat'
```
Comment (dsl96) [inline utils/consts.py]: change to non absolut path

File: utils/consts.py [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+LEFT_START, LEFT_END = face_utils.FACIAL_LANDMARKS_IDXS["left_eye"]
+RIGHT_START, RIGHT_END = face_utils.FACIAL_LANDMARKS_IDXS["right_eye"]
+PREDICTOR = dlib.shape_predictor(PREDICTOR_PATH)
```
Comment (dsl96) [inline utils/consts.py]: PREDICTOR & FACE_DETECTOR not should be in consts

File: utils/consts.py [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+FACE_DETECTOR = dlib.get_frontal_face_detector()
+
+EYE_AR_THRESH = 0.22
+MAX_BLINKS = 2
+
+ALARM_PATH = r'alarms\Alarm.mp3'  # os.path.join(os.getcwd(), os.pardir, 'alarms','Alarm.mp3')
```
Comment (dsl96) [inline utils/consts.py]: change to non absolut path

**Review by dsl96 — CHANGES_REQUESTED** [review]

File: blinked_detector.py (hunk 5 above)
Comment (israeljacob) [reply blinked_detector.py:58]: It is not a npArray

**Pull request merged**